modules/passkey/contracts/SafeWebAuthnSignerFactory.sol

```diff
@@ -2,21 +2,33 @@
 pragma solidity >=0.8.0;
 
 import {ISafeSignerFactory} from "./interfaces/ISafeSignerFactory.sol";
-import {ERC1271} from "./libraries/ERC1271.sol";
-import {P256, WebAuthn} from "./libraries/WebAuthn.sol";
-import {SafeWebAuthnSigner} from "./SafeWebAuthnSigner.sol";
+import {SafeWebAuthnSignerProxy} from "./SafeWebAuthnSignerProxy.sol";
+import {SafeWebAuthnSignerSingleton} from "./SafeWebAuthnSignerSingleton.sol";
+import {P256} from "./libraries/WebAuthn.sol";
 
 /**
- * @title WebAuthnSignerFactory
- * @dev A factory contract for creating and managing WebAuthn signers.
+ * @title SafeWebAuthnSignerFactory
+ * @dev A factory contract for creating and managing WebAuthn proxy signers.
  */
 contract SafeWebAuthnSignerFactory is ISafeSignerFactory {
+    address public immutable SINGLETON;
```

Comment:
ubernit: Since here we know **for sure** that `SINGLETON` is a `SafeWebAuthnSignerSingleton`, I wouldn't use the `address` type.

Reply:
I kept it address because the code does not use Singleton related function anywhere and only need address in assembly call. If we use `SafeWebAuthnSignerSingleton` then we need to cast it to `address` which incurs 2 gas cost.

Reply:
In theory, `SafeWebAuthnSignerSingleton` is already an address. I'm surprised that the compiler generates different code. I guess this only affects the `getSigner` call.

Reply:
also affects `isValidSignatureForSigner` and `createSigner`

Reply:
I was curious as to what 2 gas opcode instruction the compiler was adding for no reason, however, when checking locally there was no change between using address or the actual contract type:

**Before**
```
  Gas Benchmarking [@bench]
    SafeWebAuthnSignerProxy
      ⛽ deployment: 108099
      ✔ Benchmark signer deployment cost (763ms)
...
      ⛽ verification (Dummy): 14420
      ✔ Benchmark signer verification cost with Dummy verifier
```

**After**:
```
  Gas Benchmarking [@bench]
    SafeWebAuthnSignerProxy
      ⛽ deployment: 108099
      ✔ Benchmark signer deployment cost (934ms)
...
      ⛽ verification (Dummy): 14420
      ✔ Benchmark signer verification cost with Dummy verifier
```

(note that there is natural variance in the costs of actually recovering signatures with Daimo.eth and FCL)

FWIW, here is my patch

```diff
diff --git a/modules/passkey/contracts/SafeWebAuthnSignerFactory.sol b/modules/passkey/contracts/SafeWebAuthnSignerFactory.sol
index 7a6a682..3fb4dca 100644
--- a/modules/passkey/contracts/SafeWebAuthnSignerFactory.sol
+++ b/modules/passkey/contracts/SafeWebAuthnSignerFactory.sol
@@ -11,10 +11,10 @@ import {P256} from "./libraries/WebAuthn.sol";
  * @dev A factory contract for creating and managing WebAuthn proxy signers.
  */
 contract SafeWebAuthnSignerFactory is ISafeSignerFactory {
-    address public immutable SINGLETON;
+    SafeWebAuthnSignerSingleton public immutable SINGLETON;
 
     constructor() {
-        SINGLETON = address(new SafeWebAuthnSignerSingleton());
+        SINGLETON = new SafeWebAuthnSignerSingleton();
     }
 
     /**
@@ -24,7 +24,7 @@ contract SafeWebAuthnSignerFactory is ISafeSignerFactory {
         bytes32 codeHash = keccak256(
             abi.encodePacked(
                 type(SafeWebAuthnSignerProxy).creationCode,
-                uint256(uint160(SINGLETON)),
+                uint256(uint160(address(SINGLETON))),
                 x,
                 y,
                 uint256(P256.Verifiers.unwrap(verifiers))
@@ -40,7 +40,7 @@ contract SafeWebAuthnSignerFactory is ISafeSignerFactory {
         signer = getSigner(x, y, verifiers);
 
         if (_hasNoCode(signer)) {
-            SafeWebAuthnSignerProxy created = new SafeWebAuthnSignerProxy{salt: bytes32(0)}(SINGLETON, x, y, verifiers);
+            SafeWebAuthnSignerProxy created = new SafeWebAuthnSignerProxy{salt: bytes32(0)}(address(SINGLETON), x, y, verifiers);
             require(address(created) == signer);
         }
     }
@@ -55,7 +55,7 @@ contract SafeWebAuthnSignerFactory is ISafeSignerFactory {
         uint256 y,
         P256.Verifiers verifiers
     ) external view override returns (bytes4 /*magicValue*/) {
-        address singleton = SINGLETON;
+        SafeWebAuthnSignerSingleton singleton = SINGLETON;
         bytes memory data = abi.encodePacked(
             abi.encodeWithSignature("isValidSignature(bytes32,bytes)", message, signature),
             x,

```

Reply:
ohh.. i see... I was getting the prompt in vscode as below and thought prompt text is buggy but casting would incur additional 2 gas.

<img width="393" alt="Screenshot 2024-04-23 at 10 11 13" src="https://github.com/safe-global/safe-modules/assets/16358982/4f93bcfb-0d84-4891-a4d6-3ddacc7de87e">

I would cast to address as mentioned by you.

Reply:
ooh.. i see.. Vscode was giving buggy incorrect prompt when casting to address. I would follow your suggesstion.

Reply:
https://github.com/safe-global/safe-modules/pull/370/commits/f35a2eb452100fa33f5250391bea457fd980b89f